text/0300-programmatic-toolkit.md

```diff
@@ -0,0 +1,745 @@
+  source: 'toolkit' | 'user';
+  type: 'authentication' | 'cloudformation' | 'validation' ...;
+  fqn?: string;
+}
+```
+
+The toolkit might throw other exceptions.
+These are bugs and you should report them by [raising an issue](https://github.com/aws/aws-cdk/issues/new?assignees=&labels=bug%2Cneeds-triage&projects=&template=bug-report.yml&title=%28toolkit%29%3A+Untyped+Exception+%28short+description%29).
+To assist integrators with detecting the type of error, the following helper methods are available.
+Even though errors are typed, you should not rely on `instanceof` checks because it can behave unexpectedly when working with multiple copies of the same package. [TODO: insert reference link]
+
+```ts
+try {
+  await cdk.deploy(...);
+} catch (e: unknown) {
+  if (ToolkitError.isToolkitError(error)) {
+    // handle toolkit errors
+  } else if (ToolkitError.isUserError(e)) {
+    // handle user code errors
+  }
+  
+  if (ToolkitError.isCdkError(e)) {
+    // handle any cdk errors
+  }
+}
+```
+
+##### Recoverable Errors [TODO: Naming. All errors in JS are recoverable. Ruby calls this rescue/retry]
+
+  [Alternative: Do not do this at all]
+  [Alternative: Formalize a retry budget]
+
+While most errors are terminal, it's possible to recover from some.
+For example if an authentication token is expired, an integrator may prompt the user to re-authenticate via a browser login, before continuing with the action.
+With recoverable errors, the integrator can return to normal program flow after an issue has been addressed.
+
+In the [Programmatic Toolkit], recoverable errors are implemented as a special type of request.
+By default, errors are not attempted to recover.
+
+```ts
+interface RecoverableError extends IoRequest<T, U> {
+  level: 'error';
+  defaultResponse: {
+    retry: boolean; // indicates if the block should retried
+    // ... other response values
+  };
+  data: {
+    isRecoverable: true;
+    attempt: number; // the number of the current attempt, starting with 1
+  }
+}
+```
+
+When `IoHost.requestResponse()` is called with a recoverable error, the integrator may choose to return `retry: true` to indicate the block should be retried.
+A single retry is allowed, if the block fails again with the same error it will be raised as an exception.
+Retries are not guaranteed and integrators must handle the case where a requested retry is not executed.
+Recoverable errors are rare as they require special programming and not always possible.
+
+##### Synth-time Errors
+
+There is a subtle semantic difference between errors that originate from the [Programmatic Toolkit] and from a Cloud Executable, i.e. from a user's app:
+Errors from the [Programmatic Toolkit] are typically misconfigurations and maybe fixable by an integrator,
+while errors from a Cloud Executable are usually code problems and only fixable by the user.
+
+The [Programmatic Toolkit] emits all synth-time errors as recoverable errors.
```

Comment:
This surprises me. How do you expect an integrator to recover from this?

Reply:
Mostly if an integrator is also generating the code, e.g. a visual builder tool. But I think we can use try-catch for this use case for now.